compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/scene/ComposeSceneLayers.uikit.kt

```diff
@@ -0,0 +1,205 @@
+    }
+
+    fun attach(window: UIWindow, layer: UIKitComposeSceneLayer, hasViewAppeared: Boolean) {
+        val isFirstLayer = layers.isEmpty()
+
+        layers.add(layer)
+
+        view.insertSubview(layer.view, belowSubview = metalView)
+        NSLayoutConstraint.activateConstraints(
+            layer.view.layoutConstraintsToMatch(view)
+        )
+
+        if (isFirstLayer) {
+            // The content of previous layers drawn on the Metal view should be cleared and
+            // redrawn synchronously after the new layer is attached to avoid flickering.
+
+            metalView.setNeedsSynchronousDrawOnNextLayout()
+
+            window.addSubview(view)
+            NSLayoutConstraint.activateConstraints(
+                view.layoutConstraintsToMatch(window)
+            )
+            window.layoutIfNeeded()
+        }
+
+        if (hasViewAppeared) {
+            layer.sceneDidAppear()
+        }
+    }
+
+    fun detach(layer: UIKitComposeSceneLayer, hasViewAppeared: Boolean) {
+        if (hasViewAppeared) {
+            layer.sceneWillDisappear()
+        }
+
+        layers.remove(layer)
+
+        // Intercept the actions UIKitInteropTransaction from the layer
+        val transaction = layer.retrieveInteropTransaction()
+
+        if (layers.isEmpty()) {
+            // It was the last layer, remove the view and executed the actions immediately
+            view.removeFromSuperview()
+
+            transaction.actions.fastForEach { it.invoke() }
+        } else {
+            // It wasn't the last layer, pending transactions should be added to the list
+            removedLayersTransactions.add(transaction)
+
+            // Redraw content with layer removed
+            metalView.setNeedsRedraw()
+
+        }
+    }
+
+    fun viewDidAppear() {
+        layers.fastForEach {
+            it.sceneDidAppear()
+        }
+    }
+
+    fun viewWillDisappear() {
+        layers.fastForEach {
+            it.sceneWillDisappear()
+        }
+    }
+
+    /**
+     * Iterate through existing layers and merge their interop transactions to be consumed by the
+     * [MetalView], also include transactions of the layers that were removed and are not
+     * present in [layers] anymore.
+     */
+    private fun retrieveAndMergeInteropTransactions(): UIKitInteropTransaction {
+        val removedLayersTransactionsCopy = removedLayersTransactions.toList()
+        removedLayersTransactions.clear()
+
+        val transactions = layers.map { it.retrieveInteropTransaction() } + removedLayersTransactionsCopy
+        return UIKitInteropTransaction.merge(
+            transactions = transactions
+        )
+    }
+
+    private fun render(canvas: Canvas, nanoTime: Long) {
+        val composeCanvas = canvas.asComposeCanvas()
+
+        // Some layers may be removed during rendering, because recomposition will happen in the
+        // process, so we need to make a temporary copy of the list
+        layersCache.addAll(layers)
```

Comment:
what about just copying and iterating over the `layers` list?

Reply:
I don't follow, isn't it exactly what we do here?

Reply:
I mean
```
        layers.toList().fastForEach {
            it.render(composeCanvas, nanoTime)
        }
```
And we don't need `layersCache` anymore

Reply:
It's a hot path and we need to reduce allocations here. It was my request store the cache

Reply:
Ah, it's written just here. My suggestion was to reuse `CopiedList` from `CanvasLayersComposeScene.` Let's introduce helper function at least to avoid mixing rendering logic and copying stuff